src/wallet/api/wallet2_api.h

```diff
@@ -1144,6 +1220,368 @@ struct Wallet
 
     //! get bytes sent
     virtual uint64_t getBytesSent() = 0;
+
+    /**
+    * brief: getMultisigSeed - get mnemonic seed phrase for multisig wallet
+    * param: seed_offset - passphrase
+    * return: mnemonic seed phrase if succeeded, else empty string
+    * note: sets status error on fail
+    */
+    virtual std::string getMultisigSeed(const std::string &seed_offset) const = 0;
+    /**
+    * brief: getSubaddressIndex - get major and minor index of provided subaddress
+    * param: address - main- or sub-address to get the index from
+    * return: [major_index, minor_index] if succeeded
+    * note: sets status error on fail
+    */
+    virtual std::pair<std::uint32_t, std::uint32_t> getSubaddressIndex(const std::string &address) const = 0;
+    /**
+    * brief: freeze - freeze enote "so they don't appear in balance, nor are considered when creating a transaction, etc." (https://github.com/monero-project/monero/pull/5333)
+    * param: idx - index of enote in `m_transfers`
+    * param: key_image - key image of enote
+    * note: sets status error on fail
+    */
+    virtual void freeze(std::size_t idx) = 0;
+    virtual void freeze(const std::string &key_image) = 0;
+    /**
+    * brief: thaw - thaw enote that is frozen, so it appears in balance and can be spent in a transaction
+    * param: idx - index of enote in `m_transfers`
+    * param: key_image - key image of enote
+    * note: sets status error on fail
+    */
+    virtual void thaw(std::size_t idx) = 0;
+    virtual void thaw(const std::string &key_image) = 0;
+    /**
+    * brief: isFrozen - check if enote is frozen
+    * param: idx - index of enote in `m_transfers`
+    * param: key_image - key image of enote
+    * return : true if enote is frozen, else false
+    * note: sets status error on fail
+    */
+    virtual bool isFrozen(std::size_t idx) const = 0;
+    virtual bool isFrozen(const std::string &key_image) const = 0;
+    /**
+    * brief: createOneOffSubaddress - create a subaddress for given index
+    * param: account_index - major index
+    * param: address_index - minor index
+    */
+    virtual void createOneOffSubaddress(std::uint32_t account_index, std::uint32_t address_index) = 0;
+    /**
+    * brief: getWalletState - get information about the wallet
+    * return: WalletState object
+    */
+    virtual WalletState getWalletState() const = 0;
+    /**
+    * brief: rewriteWalletFile - rewrite the wallet file for wallet update
+    * param: wallet_name - name of the wallet file (should exist)
+    * param: password - wallet password
+    * note: sets status error on fail
+    */
+    virtual void rewriteWalletFile(const std::string &wallet_name, const std::string &password) = 0;
+    // QUESTION : Should we change this function from the current behavior in wallet2, so `wallet_name` is just the name of the new wallet instead of changing the `m_wallet_file` for the current wallet?
```

Comment:
I don't understand the question. Can you please elaborate?

Reply:
Sure.
The current behavior of this function in wallet2 ([src](https://github.com/monero-project/monero/blob/a1dc85c5373a30f14aaf7dcfdd95f5a7375d3623/src/wallet/wallet2.cpp#L6201)) aiui:
We provide a parameter `wallet_name` and  an outparam `new_keys_file_name`.
The first call is to `prepare_file_names(wallet_name);`, which then calls `do_prepare_file_names(wallet_name, m_keys_file, m_wallet_file, m_mms_file)`. Inside `do_prepare_file_names()` lays the problem, it changes the `m_wallet_file` of the current wallet to the new `wallet_name`.

Before going into more detail (I can give more if desired):
As I typed this I just took another look into where this `write_watch_only_wallet()` is actually used and saw it's only one place in simplewallet, where it uses `m_wallet_file` as the `wallet_name`. So even though the current behavior aiui is a mess, it's probably unimportant, and can be circumvented by removing the `wallet_name` as parameter from the API function and make it always use the current `m_wallet_file` internally. Ideally `new_keys_file_name` should not be an outparam, if you want to write a new watch only wallet with a different name than `m_wallet_file + "-watchonly.keys"`.

Let me know if things are still unclear.